src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+        {
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="ReadOnlyBuffer{T}"/> from the <see cref="T:T[]"/> offset and index.
+        /// </summary>
+        public ReadOnlyBuffer(T[] array, int offset, int length)
+        {
+            if (array == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            }
+
+            _bufferStart = new SequencePosition(array, offset | ArrayStartMask);
+            _bufferEnd = new SequencePosition(array, offset + length | ArrayEndMask);
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="ReadOnlyBuffer{T}"/> from the <see cref="ReadOnlyMemory{T}"/>.
+        /// Consumer is expected to manage lifetime of memory until  <see cref="ReadOnlyBuffer{T}"/> is not used anymore.
+        /// </summary>
+        public ReadOnlyBuffer(ReadOnlyMemory<T> readOnlyMemory)
+        {
+            ReadOnlyBufferSegment segment = new ReadOnlyBufferSegment
+            {
+                Memory = MemoryMarshal.AsMemory(readOnlyMemory)
+            };
+            _bufferStart = new SequencePosition(segment, 0 | MemoryListStartMask);
+            _bufferEnd = new SequencePosition(segment, readOnlyMemory.Length | MemoryListEndMask);
+        }
+        /// <summary>
+        /// Creates an instance of <see cref="ReadOnlyBuffer{T}"/> from the <see cref="OwnedMemory{T}"/>.
+        /// Consumer is expected to manage lifetime of memory until  <see cref="ReadOnlyBuffer{T}"/> is not used anymore.
+        /// </summary>
+        public ReadOnlyBuffer(OwnedMemory<T> ownedMemory): this(ownedMemory, 0, ownedMemory.Length)
+        {
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="ReadOnlyBuffer{T}"/> from the <see cref="OwnedMemory{T}"/>.
+        /// Consumer is expected to manage lifetime of memory until  <see cref="ReadOnlyBuffer{T}"/> is not used anymore.
+        /// </summary>
+        public ReadOnlyBuffer(OwnedMemory<T> ownedMemory, int offset, int length)
+        {
+            if (ownedMemory == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.ownedMemory);
+            }
+
+            _bufferStart = new SequencePosition(ownedMemory, offset | OwnedMemoryStartMask);
+            _bufferEnd = new SequencePosition(ownedMemory, length | OwnedMemoryEndMask);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="start"/>, and is at most <paramref name="length"/> items
+        /// </summary>
+        /// <param name="start">The index at which to begin this slice.</param>
+        /// <param name="length">The length of the slice</param>
+        public ReadOnlyBuffer<T> Slice(long start, long length)
+        {
+            SequencePosition begin = Seek(_bufferStart, _bufferEnd, start, false);
```

Comment:
Can you explain where we should call BoundsCheck before calling Seek and where we shouldn't? Like in this case, we don't call BoundsCheck.

Reply:
We call BoundsCheck when we get Position passed in.